Experimental

suresttexas00 opened the pull request:
Updates development to same version as master.

suresttexas00 commented:
Great Work!

suresttexas00 commented on src/proxy/server.py:
```diff
@@ -0,0 +1,539 @@
+        if pkid == 0x02 and self.state == 2: # Login Success - UUID & Username are sent in this packet
+            self.state = 3
+            return False
+        if pkid == self.pktCB.joingame and self.state == 3:
```
Did you get this changed later? (self.pktCB.joingame)

JasonBristol replied:
Yah those should have been cleaned up in my later commits to the PR

suresttexas00 commented on src/proxy/server.py:
```diff
@@ -0,0 +1,539 @@
+
+        if not self.client.isLocal and kill_client:  # Ben's cross-server hack
+            self.client.isLocal = True
+            self.client.send(self.pktCB.changegamestate, "ubyte|float", (1, 0))  # "end raining"
```
These two also use my old non-capped versions

suresttexas00 commented on src/proxy/server.py:
```diff
@@ -198,10 +211,14 @@ def parse(self, pkid, original):  # client - bound parse ("Server" class connect
-            if data["translate"] == "chat.type.admin":
-                return False
+            try:
+                if data["translate"] == "chat.type.admin": # Why was this being error handled?
```
This is being error handled because not all chat Json objects contain a "translate" key.

JasonBristol replied:
Just wanted to confirm, we could do

```python
if "translate" in data:
    if data["translate"] == "chat.type.admin":
        ....
```

But leveraging a try-catch may actually be faster, I will run some tests to see if checking for a key first is more efficient than handling a failure

JasonBristol replied:
It looks like using a try-catch should only be used for optimizations within say a loop, this would prevent constant lookups, but for the most part using a conditional is clearer as in this case

suresttexas00 commented on src/proxy/server.py:
```diff
@@ -122,32 +129,36 @@ def parse(self, pkid, original):  # client - bound parse ("Server" class connect
+
         if pkid == 0x01 and self.state == 2:
+            self.log.debug("Parsing.............................. 0x01 and state 2")
+            # This is throwing a malformed json, this should be a json string
```
I believe the protocol accepts simple strings too, but it should technically be "strict" json now.

JasonBristol replied:
That is what I thought too, but this seems to cause a malformed json when online-mode is enabled, might be a new 1.9 thing since I know this wasn't an issue in 1.8

suresttexas00 commented:
I don't think plugins should be importing wrapper's modules.. Now I see what some of the to-do was about in our discussion yesterday...
The problem with this (besides these things getting changed) is that
1) plugin devs should not need to access these items directly.  We have api methods for it.  Additional methods can be added at the devs request... direct access should be a temporary stop-gap measure until an update with the authorized method in made.
2) I have concerns that just importing, say, base.proxy, does not mean you are getting the live running instance of proxy (the one that is bound to the server's port).  Any methods you run would be init'ed but not have running values corresponding to the live server... just a thought; I could be wrong.
3) Since Plugins are imported by wrapper, which already incorporates the wrapper files, you are running an import that re-imports part of your self.

A properly coded Plugin should be using the documented API exclusively... exceptions begin made for bleeding edge plugins that should be awaiting the same functionality from a new wrapper release.

JasonBristol commented:
All of your points are from the perspective of the way wrapper is today. We are looking to refactor the entire plugin implementation and how plugins are pipelined through wrapper.

I agree with point 1, but this is dependent on a refactor and extension of our API. The reason why we are importing modules directly is because the API doesn't allow us to do what we want always.

I am not sure what you mean with point 2, I wouldn't be importing proxy.base with the intention of getting an instance of a proxy object, I would be doing it to sniff packet content or call proxy methods directly. If I want the proxy object, I would need to create a getter method in proxy.base.

I agree with point 3, the issue isn't with the idea of importing though, the issue is the way wrapper works. It passes a wrapper object into each plugin, instead of allowing the user to import the wrapper base and leverage getters and setters to interact directly. We are quite litterally bypassing the very intention of python modules (i.e. importing etc).

As the name suggest a python module is a modular set of code that should be dynamic enough for me to import __anywhere__ and leverage it. This is actually how we want plugins to work, not the way they are today.

suresttexas00 commented:
Re: point 2- The plugin methods won't work properly, nor will you be able to sniff packets, etc, without the proxy object (the active connection).  The methods will not work.

Re point 3 - You are treating this as though the plugins are the program and wrapper is the set of modules you get to import and play with... The opposite is the case - Wrapper is the program and plugins are just the imported modules.  This is going to get the cart before the horse... and you are going to end up something that is not nearly what wrapper is.

We are supposed to be refactoring (with the end goal of making the code less buggy), not creating whole new paradigms.  The end product should function exactly as it always has, per the current documentation... Old plugins must function as they did before.  If the standard API lacks some desired functionality, then that needs to be added to the API, not make it easier to access all the private wrapper code (even if it is not "technically" so private).

If you are wanting more than just a mere refactor, perhaps also a separate re-write branch, unrelated to wrapper, in start to finish py3, and not subject to any proprietary code or copyright issues may be warranted?

JasonBristol commented:
Re: point 2: Again, with what we have today you are right, but this is what we need to fix

Re: point 3: That is not what I am suggesting, to python all modules are equal, we should embrace this to offer, not __mandate__, plugin devs more power over their wrapper. You don't have to use wrapper this way, but we shouldn't make it impossible for people to do this.

Re: point 3: You might be right, I am starting to think putting effort into this is going to be a missed opportunity for building a new wrapper. For my use case, what we have now is nice, but I am not willing to crystalize what we have as the end state. There is a whole lot more we can allow wrapper to do and I think we are either to afraid to do it, or not willing to adjust our definition of wrapper to make it a more powerful tool for server owners.

suresttexas00 commented:
People in this community use it both ways (as a simple jar wrapper and as a plugin insertion tool).

I don't have a problem with making wrapper more flexible **_as long as_** you don't break the current API...

I am also concerned about the investment/return ratios that such and effort carries... creating an interface out of the present wrapper to support working with wrappers own internal methods versus just using the provided API seems cost prohibitive in terms of the benefit... which is mostly just to have what should be a redundant methods or only used by bleeding-edge plugins until the next API comes out. That cost of making wrapper work in a new way would be better spent on a fresh version.

However, that is mind boggling that proxy takes 50% of the code base.  If you went with a new build, I would suggest a name "MCProxy".  For those of us for whom server customization is the primary focus, building a really state of the art proxy system would really do what we are asking of wrapper.

suresttexas00 committed:
Update master branch to lastest version
Primary reasons for update:
- Regular update prior to further new Wrapper development

Api improvements:
- Updates to NBT.py for various parts of Mincraft 1.12 and upcoming 1.13.
- version checking added to proxy/utils/encryption.py cryptography import.
- Make the packet flush times configurable since different computers and setups may have different ideal flushing intervals.
- allow different IP's for hubs (able to use 192.168.x.x, for instance).
- adds DiscordRelay plugin by @PurePi
- api.helpers.get_int - accepts possible booleans
- add Regions suite to stable plugins.
- Add base API property `wrapper_version` to allow plugins to inspect Wrapper's version.
- improved name changes to include offline hubs being able to update names too!
  : whitelist warning : Names are still whitelisted by offline name. An
   automatic name change will cause player to not be whitelisted (you will
   have to `whitelist add` the new name).
- Make wrapper current through snapshot 18w14b:
    - add protocol 368 as PROTOCOL_PRE_RELEASE
    - add PROTOCOL_PRE_RELEASE packets for CB and SB.
    - Fix slot parsing, which has changed with new snapshots.
    **TEMPORARILY BROKE / not implemented yet** - server.autoCompletes
     (parse_cb.py) event is not compliant for sending/modding the new packet.

Bugfixes:
- name changes were not actually working (old name persisted).
- Fixed wrapper's player.interact use_item event that has been broken
 ever since wrapper stopped using the old string-keys packet read() format.
- Fix error that causes wrapper to think snapshots are pre-netty.
- Fix old Py3 Errors in NBT things
- Harden up wrapper's handling of disconnected players.
- Patch resource imports that may not work on Windows.
- Fix wrapper client inventory bugs (inconsistent use of None versus {"id" = -1}
 in code.
- Various regions plugin bugfixes and improvements.
- Fix bug in api.backups.enableBackups.
- fix bug in portals.py
- fix bug clientconnection.py line 1074 (not enough args for log formatting).
- fix bug in parse_sb.py and parse_cb.py player objects, usually caused by player disconnection.
- Possibly fixed bug where we may not be shutting down client sockets when they abort,
 causing the os system file limits to be exceeded after a certain amount of time.
- fix error in backups.py.  It still referenced the "payload" argument from
 the original 'timer.second' event that used to drive it.
- bug fix packet.py line 645 that calls close_server method (which did not
 formerly exist for clientconnection.py).  added a `close_server` method to
 clientconnection.py to wrap cleintconnection's `_close_server_instance` method.
- Bugfix remove Python3 style function annotations.  Use comment style annoations instead.
- fix errors with occasional problems in playerposition and look type packets that kill the client connection.
- Fix documentation errors in block placement, digging, and interaction events.
- removed debug print statements in parse_sb.py
- Restore interact event's ability to parse the placement coords of buckets.

Refactors and code improvements:
- Change proxy.base.pollserver() to return values instead of modding wrapper's
 values directly.  This makes it a truly public function, not an internal method.
 Provides for future ability to poll other servers besides the locally wrapped one.
- Tie entitycontrol abort signal to wrapper/proxy abort signals, eliminate its' personal abort signal.
- refactor wrapper halt signal from halt to haltsig.
- Cleanup disconnection code some.
- Mark blockForEvent code as deprecated for removal by wrapper 1.0 final or
 1.1 rc.
- Only call Wrapper Alerts if enabled.- Possibly fixed bug where we may not be shutting down client sockets when they abort,
 causing the os system file limits to be exceeded after a certain amount of time.
- Remove on-second timer from backups.
- Change event coding to use a single thread for processing non-blocking events, utilizing a queue.
- Backups timer uses wrapper storage to store it's backup timer between restarts.
- Changed existing plugins that use timer.second to implement their own timers.
- Remove player.interact event out of the block placement code because there
 really is no way to tell if the client is interacting or not based on inventory.
 clicking on a chest with no item or clicking on a chest with a block in hand still
 opens the chest, for example.
- added a player object "player" to event payloads that did not have it.  This is
 only a cosmetic change in the API (and a speed optimization) because the event
 code already added "player" objects to payloads missing the player object,  This
 also corrects the documentation that did not list player as a payload.  Retained
 the "playername" payload in these events, for backwards compatibility.
- optimize the event processor thread timer.
- Improve plugin event code some.
- use collections.deque to decrease pop TimeComplexity from O(k) to O(1) in
 event handling code and packet.py processing (https://wiki.python.org/moin/TimeComplexity).

KNOWN BUG (which exists in previous release version too):
- does not work with compression disabled.

suresttexas00 committed:
Update development
